src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/address_union.rs

```diff
@@ -0,0 +1,449 @@
+                            NetworkID::Mainnet
+                        );
+                    )+
+
+                    $(
+                        assert_eq!(
+                            [< $union_name:snake _network_id >](&[< new_ $union_name:snake _sample_ $variant_name:snake _stokenet_other >]()),
+                            NetworkID::Stokenet
+                        );
+                    )+
+                }
+
+                #[test]
+                fn map_to_network() {
+
+                    SUT::sample_values_all().into_iter().for_each(|a| {
+                        assert_eq!(
+                            [< $union_name:snake _map_to_network>](&a, NetworkID::Stokenet).network_id(),
+                            NetworkID::Stokenet
+                        );
+                    });
+
+                    SUT::sample_values_all().into_iter().for_each(|a| {
+                        assert_eq!(
+                            [< $union_name:snake _map_to_network>](&a, NetworkID::Mainnet).network_id(),
+                            NetworkID::Mainnet
+                        );
+                    });
+                }
+
+                #[test]
+                fn format_address() {
+                    let test = |a: SUT| {
+                        assert_eq!([< $union_name:snake _formatted>](&a, AddressFormat::Full), a.to_string());
+                        assert_ne!([< $union_name:snake _formatted>](&a, AddressFormat::Default), a.to_string());
+                    };
+                    SUT::sample_values_all().into_iter().for_each(test);
+                }
+
+            }
+
+            $(
+                impl From<$variant_type> for $union_name {
+                    fn from(value: $variant_type) -> Self {
+                        Self::$variant_name(value)
+                    }
+                }
+            )+
+
+            impl IntoScryptoAddress for  $union_name {
+                fn scrypto(&self) -> ScryptoGlobalAddress {
+                    match self {
+                        $(
+                            Self::$variant_name(address) => address.scrypto(),
+                        )+
+                    }
+                }
+
+                /// Returns the [`NetworkID`]
+                fn network_id(&self) -> NetworkID {
+                    match self {
+                        $(
+                            Self::$variant_name(address) => address.network_id(),
+                        )+
+                    }
+                }
+            }
+
+            impl $union_name {
+                pub fn new_from_bech32(s: &str) -> Result<Self> {
+                    let e = CommonError::FailedToDecodeAddressFromBech32 { bad_value: s.to_string() };
+                    Err(e)
+                    $(
+                        .or($variant_type::from_str(s).map(Self::from))
```

Comment:
also pretty nice, I start with an `Err` and then I can "chain" `.or(Result)` for all variants :)